openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+    {
+        typename BufferPoolT::reference buffer = mBufferPool->local();
+        typename StencilPoolT::reference stencil = mStencilPool->local();
+        const ValueT isoValue = mIsoValue, above = mAboveSign*std::numeric_limits<ValueT>::max();//local copy
+        const ValueT h = mAboveSign*static_cast<ValueT>(mGrid1->voxelSize()[0]);//Voxel size
+        size_t boundaryCount = 0;
+        for (auto leafIter = r.begin(); leafIter; ++leafIter) {
+            ValueT* sdf = leafIter.buffer(1).data();
+            for (auto voxelIter = leafIter->beginValueAll(); voxelIter; ++voxelIter) {
+                const ValueT value = *voxelIter;
+                const bool isAbove = value > isoValue;
+                if (!voxelIter.isValueOn()) {// inactive voxels
+                  sdf[voxelIter.pos()] = isAbove ? above : -above;
+                } else {// active voxels
+                  const Coord ijk = voxelIter.getCoord();
+                  stencil.moveTo(ijk, value);
+                  const auto mask = stencil.intersectionMask( isoValue );
+                  if (mask.none()) {// most common case
+                    buffer.push_back(ijk);
+                    sdf[voxelIter.pos()] = isAbove ? above : -above;
+                  } else {// compute distance to iso-surface
+                    ++boundaryCount;
+                    const ValueT delta = value - isoValue;//offset relative to iso-value
+                    if (math::isApproxZero(delta)) {//voxel is on the iso-surface
+                      sdf[voxelIter.pos()] = 0;
+                    } else {//voxel is neighboring the iso-surface
+                      ValueT sum = 0;
+                      for (int i=0; i<6;) {
+                        ValueT d = std::numeric_limits<ValueT>::max(), d2;
+                        if (mask.test(i++)) d = math::Abs(delta/(value-stencil.getValue(i)));
+                        if (mask.test(i++)) {
+                          d2 = math::Abs(delta/(value-stencil.getValue(i)));
+                          if (d2 < d) d = d2;
+                        }
+                        if (d < std::numeric_limits<ValueT>::max()) sum += 1/(d*d);
+                      }
+                      sdf[voxelIter.pos()] = isAbove ? h / math::Sqrt(sum) : -h / math::Sqrt(sum);
+                    }// voxel is neighboring the iso-surface
+                  }// intersecting voxels
+                }// active voxels
+            }// loop over voxels
+        }// loop over leaf nodes
+        mParent->mBoundaryCount += boundaryCount;//reduces pressure on the atomic
+    }// operator(const LeafRange& r)
+    template<typename RootOrInternalNodeT>
+    void operator()(const RootOrInternalNodeT& node) const
+    {
+        const ValueT isoValue = mIsoValue, above = mAboveSign*std::numeric_limits<ValueT>::max();
+        for (auto it = node.cbeginValueAll(); it; ++it) {
+          ValueT& v = const_cast<ValueT&>(*it);
+          v = v > isoValue ? above : -above;
```

Comment:
assigning to the result of a const_cast is undefined - is there a way to make this non-const? The NodeManager seems to iterate on a non-const tree, so perhaps the argument to this function can simply be non-const too?

Reply:
I'm not sure that applies here since the value that the reference is pointing to in the leaf-node is not declared as const. Changing it is tricky since the ValueAccessor (or tree) does not have any access methods that return a non-const reference. Only alternative I can think of is to get a const pointer to the buffer and cast that but I'm actually not convinced this is an issue to begin with

Reply:
I believe you are correct, I was thrown off by the fact this operator takes a const node. So my follow up question is, why do these operators accept const nodes? Can we simply remove the const from the input node argument?

Reply:
strictly you're right that they could take non-const nodes but it wouldn't change anything since the iterator still returns a const reference (a design limitation coming from the fact that LeafNode<bool> is template specialized with bit-arrays). I guess it would make the function signature more "honest". I'll change it ...

Reply:
tried this and the compiler gets confused since it cannot make the correct distinction between 

template<typename RootOrInternalNodeT>
    void operator()(RootOrInternalNodeT& node) const

and

void operator()(const LeafRange& r) const

Reply:
Yes, so long as you don't need a const Leaf/NodeManager you should also be able to remove the const there too!